Add documentation for Ractor

zverok opened the pull request:
Currently, ractors documented only by [doc/ractor.md](https://docs.ruby-lang.org/en/master/doc/ractor_md.html), which has a flavor of a design/discussion document. I wanted to establish some _base_ documentation, which will allow to understand and use the concept immediately. Currently, Ractor class [has no docs at all](https://docs.ruby-lang.org/en/master/Ractor.html) -- it is partially due to the fact that `ractor.rb` is not included in the `.document` list for RDoc, but even so, per-method [docs in ractor.rb](https://github.com/ruby/ruby/blob/master/ractor.rb) seem kinda sparse and chaotic to me.

I've completely rewritten the class docs. They are now somewhat duplicating `doc/ractor.md`, but from a different perspective.

Test rendering of the docs on my personal site: [Ractor.html](https://zverok.github.io/ruby-rdoc/Ractor.html)

@ko1 @marcandre @eregon Can you please review it?

And one question I couldn't clarify by myself: can somebody please show an example of the code throwing `Ractor::UnsafeError` and maybe provide some brief explanation about it?.. From the sources I kinda get the idea, but my knowledge is not deep enough.

marcandre commented:
I'll gladly review this tomorrow 👍

ko1 commented on ractor.rb:
```diff
@@ -153,26 +512,60 @@ class RemoteError
   end
 
   # Closes the incoming port and returns its previous state.
+  # All further attempts to Ractor.receive in the ractor, and #send to the ractor
```
I think it is valuable that incoming/outgoing ports are closed automatically when the ractor terminates.

more details:
on closed incoming port, `Ractor#send` will raise immediately.
But `Ractor.receive` doesn't raise an error until there are sent messages in incoming queue.
if the queue is empty, `Ractor.receive` raises.

outgoing port will be closed after the last yielded message (ractor's block value or exception which terminates the ractor) is taken.

zverok replied:
I too believe it is important, and explain this in docs of `#send` and and `#take`, I believe it belongs there, not to the method for explicitly closing ports, don't you think?..

> But `Ractor.receive` doesn't raise an error until there are sent messages in incoming queue.

Hm, in my experiment it raises immediately on closed incoming:

```ruby
Ractor.new { 
  close_incoming
  receive 
}
# #<Thread:0x00005652ebd41ae0 run> terminated with exception (report_on_exception is true):
# <internal:ractor>:98:in `receive': The incoming port is already closed (Ractor::ClosedError)
```

ko1 replied:
> I too believe it is important, and explain this in docs of #send and and #take, I believe it belongs there, not to the method for explicitly closing ports, don't you think?..

Ah... I think automatic closing should be also explained here (reference to the details is enough).

> Hm, in my experiment it raises immediately on closed incoming:

```ruby
r = Ractor.new Ractor.current do |main|
  main << true
  p receive #=> 1 (closed, but remained)
  p receive #=> 2 (closed, but remained)
  p receive #=> The incoming port is already closed (Ractor::ClosedError)
end

r << 1
r << 2
Ractor.receive
r.close_incoming
r.take
```

ko1 commented on ractor.rb:
```diff
@@ -153,26 +512,60 @@ class RemoteError
+  # Checks if the object is shareable by ractors.
+  #
+  #     Ractor.shareable?(1)        #=> true -- numbers and other immutable basic values are
+  #     Ractor.shareable?('foo')    #=> true or false, depending on whether you have freeze_string_literals: true
```
ditto -> String#+@

ko1 reviewed (commented):
Thank you for your documentation.

I commented, and the biggest problems are:

* should not use `sleep(0.1)` here.
* should not mention about GVL here.

eregon commented on ractor.c:
```diff
@@ -1678,6 +1678,77 @@ ractor_moved_missing(int argc, VALUE *argv, VALUE self)
+ *  Raised when an attempt is made to take something from the Ractor's outgoing port,
+ *  but it is closed with Ractor#close_outgoing, or to send something to Ractor's
+ *  incoming port, and it is closed with Ractor#close_incoming, or an attempt to
+ *  send/take something with ractor which was already terminated.
```
`StopIteration`*

zverok replied:
Good point! Updated the docs.

zverok commented on ractor.rb:
```diff
@@ -1,34 +1,183 @@
+#
+#     server = Ractor.new do
+#       puts "Server starts: #{self.inspect}"
+#       "Server sends: ping"
```
Typo. Fixed (both the script and the output)

zverok commented on ractor.rb:
```diff
@@ -1,34 +1,183 @@
+#
+#     Ractor.shareable?(1)        #=> true -- numbers and other immutable basic values are
+#     Ractor.shareable?('foo')    #=> true or false, depending on whether you have freeze_string_literals: true
+#     Ractor.shareable?(-'foo')   #=> true
```
OK, fixed. I tried to write an example in a way that would also work if somebody inserts it into the file with `# frozen_string_literals: true`, but you are right, that's obscuring the point.

zverok commented on ractor.rb:
```diff
@@ -1,34 +1,183 @@
+# When a shareable object is sent (via #send or Ractor.yield), no additional processing happens,
+# and it just becomes usable by both ractors. When an unshareable object is sent, it can be
+# either _copied_ or _moved_. The first is the default, and it makes the object's full copy by
+# dumping and loading it with Marshal.
```
My bad. Checked the implementation, updated the docs.

zverok commented on ractor.rb:
```diff
@@ -38,38 +187,75 @@ def self.new(*args, name: nil, &block)
 
+  # Returns total count of Ractors currently running.
+  #
+  #    Ractor.count                   #=> 1
```
I believe that the way it currently shows the point better (save for `sleep`, which I'll address below): with `Ractor.count` inside other ractor it is not clear: whether it would be `2` only _inside_ the second ractor, or throughout the system, while the second ractor is alive.

WormerKing commented:
sorry. just started using new github and i don't know where i clicked

14 Ara 2020 Pzt 20:04 tarihinde Victor Shepelev <notifications@github.com>
şunu yazdı:

> *@zverok* commented on this pull request.
> ------------------------------
>
> In ractor.rb
> <https://github.com/ruby/ruby/pull/3895#discussion_r542549701>:
>
> > +# check this, and Ractor.make_shareable tries to make object shareable if it is not.
> +#
> +#     Ractor.shareable?(1)        #=> true -- numbers and other immutable basic values are
> +#     Ractor.shareable?('foo')    #=> true or false, depending on whether you have freeze_string_literals: true
> +#     Ractor.shareable?(-'foo')   #=> true
> +#     Ractor.shareable?(+'foo')   #=> false
> +#
> +#     str = +'foo'
> +#     str.frozen?                 #=> false
> +#     Ractor.make_shareable(str)
> +#     str.frozen?                 #=> true
> +#
> +# When a shareable object is sent (via #send or Ractor.yield), no additional processing happens,
> +# and it just becomes usable by both ractors. When an unshareable object is sent, it can be
> +# either _copied_ or _moved_. The first is the default, and it makes the object's full copy by
> +# dumping and loading it with Marshal.
>
> My bad. Checked the implementation, updated the docs.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ruby/ruby/pull/3895#discussion_r542549701>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APFTFTWTJLLPR2NMIABKKSDSUZASXANCNFSM4UZWHG6Q>
> .
>

WormerKing commented:
Ruby is a programming language that I love very much and I am trying to do
projects to develop this programming language. But ruby ​​is a programming
language that I am trying to learn new so I cannot contribute to its
development for now.

14 Ara 2020 Pzt 20:12 tarihinde Victor Shepelev <notifications@github.com>
şunu yazdı:

> *@zverok* commented on this pull request.
> ------------------------------
>
> In ractor.rb
> <https://github.com/ruby/ruby/pull/3895#discussion_r542556464>:
>
> > @@ -153,26 +512,60 @@ class RemoteError
>    end
>
>    # Closes the incoming port and returns its previous state.
> +  # All further attempts to Ractor.receive in the ractor, and #send to the ractor
>
> I too believe it is important, and explain this in docs of #send and and
> #take, I believe it belongs there, not to the method for explicitly
> closing ports, don't you think?..
>
> But Ractor.receive doesn't raise an error until there are sent messages
> in incoming queue.
>
> Hm, in my experiment it raises immediately on closed incoming:
>
> Ractor.new {
>   close_incoming
>   receive }# #<Thread:0x00005652ebd41ae0 run> terminated with exception (report_on_exception is true):# <internal:ractor>:98:in `receive': The incoming port is already closed (Ractor::ClosedError)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ruby/ruby/pull/3895#discussion_r542556464>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APFTFTVNSY3YYQ4EHYUTM3TSUZBQJANCNFSM4UZWHG6Q>
> .
>

zverok commented:
@ko1 I addressed (or replied) most of your comments.
The one problem I have is with `sleep`. I use it as a generic "wait till all currently running ractors will do what they can". Using `Ractor#take` instead is undesirable in lot of the examples, starting with one of the base ones:
```ruby
server = Ractor.new do
  puts "Server starts: #{self.inspect}"
  puts "Server sends: ping"
  Ractor.yield 'ping'                       # The server doesn't know the receiver and sends to whoever interested
  received = Ractor.receive                 # The server doesn't know the sender and receives from whoever sent
  puts "Server received: #{received}"
end

client = Ractor.new(server) do |srv|        # The server is sent inside client, and available as srv
  puts "Client starts: #{self.inspect}"
  received = srv.take                       # The Client takes a message specifically from the server
  puts "Client received from " \
       "#{srv.inspect}: #{received}"
  puts "Client sends to " \
       "#{srv.inspect}: pong"
  srv.send 'pong'                           # The client sends a message specifically to the server
end
```
Who exactly should I `take` from to definitely finalize the script successfully after all the communication have happened, and not disrupt their communication (e.g., not receive in the main ractor server's `"ping"`, instead of the client)?

What I actually need for showcasing examples is some generic `Ractor.tick`, or `Ractor.dispatch_all` with the meaning "let them do what they could", or, maybe `Ractor#wait` with the meaning "wait till Ractor finishes, but not take its value, and not raise ClosedError if it already was", e.g. some more direct analog of `Thread#join`. 

Though, in this example neither `#take` nor (imaginary) `#wait` will not do, it requires `Ractor.tick`:
```ruby
r = Ractor.new {
  puts "Before first receive"
  v1 = Ractor.receive
  puts "Received: #{v1}"
  v2 = receive
  puts "Received: #{v2}"
}
sleep(0.1) # <== HERE
puts "Still not received"
r.send('message1')
sleep(0.1)
puts "Still received only one"
r.send('message2')
sleep(0.1)
```
At the place which I've marked with "HERE", `sleep` demonstrates "Ractor has a chance to run, but it did not yet". `Ractor#take` will not do here, and I am not sure what else will.

So, I am not sure how to proceed, do you have some ideas? 

In the absence of abstract `tick`, `sleep` is the closest thing I could find. Maybe one option is to define at the beginning of explanations 
```ruby
def tick
  sleep(0.1)
end
```
...and explain "Will call this toy method to give all running ractors a chance to properly dispatch what they are doing currently". Would it be better?

WormerKing commented:
Thank you!

14 Ara 2020 Pzt 21:17 tarihinde Victor Shepelev <notifications@github.com>
şunu yazdı:

> @ko1 <https://github.com/ko1> I addressed (or replied) most of your
> comments.
> The one problem I have is with sleep. I use it as a generic "wait till
> all currently running ractors will do what they can". Using Ractor#take
> instead is undesirable in lot of the examples, starting with one of the
> base ones:
>
> server = Ractor.new do
>   puts "Server starts: #{self.inspect}"
>   puts "Server sends: ping"
>   Ractor.yield 'ping'                       # The server doesn't know the receiver and sends to whoever interested
>   received = Ractor.receive                 # The server doesn't know the sender and receives from whoever sent
>   puts "Server received: #{received}"end
> client = Ractor.new(server) do |srv|        # The server is sent inside client, and available as srv
>   puts "Client starts: #{self.inspect}"
>   received = srv.take                       # The Client takes a message specifically from the server
>   puts "Client received from " \
>        "#{srv.inspect}: #{received}"
>   puts "Client sends to " \
>        "#{srv.inspect}: pong"
>   srv.send 'pong'                           # The client sends a message specifically to the serverend
>
> Who exactly should I take from to definitely finalize the script
> successfully after all the communication have happened, and not disrupt
> their communication (e.g., not receive in the main ractor server's "ping",
> instead of the client)?
>
> What I actually need for showcasing examples is some generic Ractor.tick,
> or Ractor.dispatch_all with the meaning "let them do what they could",
> or, maybe Ractor#wait with the meaning "wait till Ractor finishes, but
> not take its value, and not raise ClosedError if it already was", e.g. some
> more direct analog of Thread#join.
>
> Though, in this example neither #take nor (imaginary) #wait will not do,
> it requires Ractor.tick:
>
> r = Ractor.new {
>   puts "Before first receive"
>   v1 = Ractor.receive
>   puts "Received: #{v1}"
>   v2 = receive
>   puts "Received: #{v2}"}sleep(0.1) # <== HEREputs "Still not received"r.send('message1')sleep(0.1)puts "Still received only one"r.send('message2')sleep(0.1)
>
> At the place which I've marked with "HERE", sleep demonstrates "Ractor
> has a chance to run, but it did not yet". Ractor#take will not do here,
> and I am not sure what else will.
>
> So, I am not sure how to proceed, do you have some ideas?
>
> In the absence of abstract tick, sleep is the closest thing I could find.
> Maybe one option is to define at the beginning of explanations
>
> def tick
>   sleep(0.1)end
>
> ...and explain "Will call this toy method to give all running ractors a
> chance to properly dispatch what they are doing currently". Would it be
> better?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ruby/ruby/pull/3895#issuecomment-744619760>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/APFTFTXBKNPPUEHSWMOW5DTSUZJB7ANCNFSM4UZWHG6Q>
> .
>

marcandre commented on ractor.rb:
```diff
@@ -83,7 +266,51 @@ def self.select(*ractors, yield_value: yield_unspecified = true, move: false)
-  # Receive an incoming message from Ractor's incoming queue.
+  # Receive an incoming message from the current Ractor's incoming port's queue, which was
+  # sent there by #send.
+  #
```
It wasn't clear to me that `Ractor.receive` and `Ractor#receive` have the same effect. Should we mention it?

marcandre commented on ractor.rb:
```diff
@@ -102,39 +329,174 @@ class << self
+  #    puts "Sent successfully"
+  #    # Prints: "Sent successfully" immediately
+  #
+  # If close_incoming was called on the ractor, the method raises Ractor::ClosedError.
```
I'd start with a ractor which already finished, which should be more common

marcandre commented on ractor.rb:
```diff
@@ -102,39 +329,174 @@ class << self
+  #
+  # This prints:
+  #
+  #    Receieved: message
```
typo: Received

marcandre commented on ractor.rb:
```diff
@@ -102,39 +329,174 @@ class << self
+  #    puts r.take
+  #    # Prints: "Hello from ractor"
+  #
+  # The method is blocking, and will return only when somebody will consume the
```
^will consume^consumes

marcandre commented on ractor.rb:
```diff
@@ -102,39 +329,174 @@ class << self
+  #    # Ractor::ClosedError (The outgoing-port is already closed)
+  #    # The error would be raised immediately, not when ractor will try to receive
+  #
+  # If the exception have happened in the Ractor, it is propagated on take as a Ractor::RemoteError.
```
"If the exception have happened" => "If an uncaught exception is raised"

marcandre commented on ractor.rb:
```diff
@@ -102,39 +329,174 @@ class << self
+  #     r.take
+  #   rescue => e
+  #     p e         # => #<Ractor::RemoteError: thrown by remote Ractor.>
+  #     p e.ractor  # => #<Ractor:#2 test.rb:1 terminated>
```
`p e.ractor == r # => true` seems more clear

marcandre commented on ractor.rb:
```diff
@@ -1,34 +1,180 @@
+#
+# Notice that even +inspect+ (and more basic methods like <tt>__id__</tt>) is inaccessible
+# on a moved object.
+#
```
We need a section for other shareable objects: `Modules, Classes, Ractors`.

zverok replied:
Do we have a full list somewhere (in code?) The `doc/ractor.md` is a bit sparse on that matter.

marcandre replied:
Code is not clear to me, but I believe that's it.
We should mention that thede are shareable but can not access instance variables from outside the main Ractor.

zverok replied:
Added. Also, I noticed I forgot to check and explain constant access rules, fixed it, too :)

zverok replied:
Funny that access to non-frozen constant value throws `NameError` and access to class instance vars throws `RuntimeError`, and not something Ractor-specific.

marcandre reviewed (approved):
Good work, lots of doc here 💪

zverok commented on ractor.c:
```diff
@@ -1678,6 +1678,91 @@ ractor_moved_missing(int argc, VALUE *argv, VALUE self)
+ *
+ *  Raised on an attempt to access an object which was moved in Ractor#send or Ractor.yield.
+ *
+ *     r = Ractor.new { receive }
```
I am not sure which is ~worse~ better. From the one point, avoiding `recieve` will underline that it is `send` which causes moving. From the other, on example this short, "sending to something that can't receive" might feel confusing.
But changing :)

zverok commented on ractor.c:
```diff
@@ -1678,6 +1678,91 @@ ractor_moved_missing(int argc, VALUE *argv, VALUE self)
+ *
+ *  A special object which replaces any value that was moved to another ractor in Ractor#send
+ *  or Ractor.yield. Any attempt to access the object results in Ractor::MovedError.
+ *
```
Good point. I was actually surprised by this when playing with examples.
BTW, do you know why the object is so severely limited, would it break something if it would have not frozen class (like even `nil` has), nice `#inspect` an so on?

marcandre replied:
I believe the intent is to enforce that it is a programming error to refer to that value, but I'm not sure. It was clever of you to do `MovedObject === obj`, I didn't think of that.
Note that refinements work even though the class is frozen...

marcandre replied:
I'll open an issue, because it might be helpful to actually have some basic introspection info like `inspect`.

zverok replied:
BTW, it seems `p` bypasses `#inspect` redefinition (and `#to_s` redefinition to format the message!):
```ruby
ary = []
Ractor.new{ sleep }.send(ary, move: true)
ary.inspect
# Ractor::MovedError (can not send any methods to a moved object)
p ary
# NoMethodError (undefined method `inspect' for #<Ractor::MovedObject:0x00005598b4093dd0>)
```

marcandre replied:
Yes, I believe this is standard refinement behavior; methods defined outside the `using` can not see refined methods. It can be confusing because some syntax sugar does, e.g. `"...#{foo}..."` is like syntax sugar for `"...#{foo.to_s}..."`, but that's all I can think of. `[foo].join` will not see a refined `to_s`

zverok commented on ractor.rb:
```diff
@@ -1,34 +1,180 @@
+#     Client received from #<Ractor:#2 rac.rb:1 blocking>: ping
+#     Client sends to #<Ractor:#2 rac.rb:1 blocking>: pong
+#     Server received: pong
+#
```
Not sure. The introduction is long enough already, and particular methods' docs state that clearly, I don't think one's understanding of the base concept would be enhanced by this.

zverok commented on ractor.rb:
```diff
@@ -1,34 +1,180 @@
+#
+#     str = 'foo'
+#     str.frozen?                 #=> false
+#     Ractor.make_shareable(str)
```
Yeah, makes sense. I provide more detailed example in the method's docs, but it should be here too.

marcandre commented:
> The one problem I have is with `sleep`. I use it as a generic "wait till all currently running ractors will do what they can". Using `Ractor#take` instead is undesirable in lot of the examples

There's no perfect answer. `sleep` is not completely deterministic and should be avoided in actual code. OTOH it's still not clear to me if `take` should be encouraged for standard communication between ractors. I have the impression that it is best reserved for Ractor process monitoring, but I may be mistaken.
For me it's clear that `sleep` is just an example and shouldn't be used in production.

zverok commented on ractor.rb:
```diff
@@ -1,34 +1,180 @@
+# When a shareable object is sent (via #send or Ractor.yield), no additional processing happens,
+# and it just becomes usable by both ractors. When an unshareable object is sent, it can be
+# either _copied_ or _moved_. The first is the default, and it makes the object's full copy by
+# deep cloning of its structure.
```
Yeah, makes sense. I changed the whole explanation/example to:

When a shareable object is sent (via #send or Ractor.yield), no additional processing happens,
and it just becomes usable by both ractors. When an unshareable object is sent, it can be
either _copied_ or _moved_. The first is the default, and it makes the object's full copy by
deep cloning of non-shareable parts of its structure.

```ruby
data = ['foo', 'bar'.freeze]
r = Ractor.new{
 data2 = Ractor.receive
 puts "In ractor: #{data2.object_id}, #{data2[0].object_id}, #{data2[1].object_id}"
}
r.send(data)
sleep(0.1)
puts "Outside  : #{data.object_id}, #{data[0].object_id}, #{data[1].object_id}"
```

This will output:

```
In ractor: 340, 360, 320
Outside  : 380, 400, 320
```

(Note that object id of both array and non-frozen string inside array have changed inside
the ractor, showing it is different objects. But the second array's element, which is a
shareable frozen string, has the same object_id.)

marcandre replied:
I like it

zverok commented:
@marcandre Thanks for the review! I tried to address your comments.

ko1 commented:
As @marcandre said, `sleep 0.1` is not deterministic (slow computer can not run expectedly) and is should not be used.
If documentation uses it, people can use it.

Instead of sleep, we need to use appropriate communication (synchronization) mechanism.
 
> @ko1 I addressed (or replied) most of your comments.
> The one problem I have is with `sleep`. I use it as a generic "wait till all currently running ractors will do what they can". Using `Ractor#take` instead is undesirable in lot of the examples, starting with one of the base ones:
> 
> ```ruby
> server = Ractor.new do
>   puts "Server starts: #{self.inspect}"
>   puts "Server sends: ping"
>   Ractor.yield 'ping'                       # The server doesn't know the receiver and sends to whoever interested
>   received = Ractor.receive                 # The server doesn't know the sender and receives from whoever sent
>   puts "Server received: #{received}"
> end
> 
> client = Ractor.new(server) do |srv|        # The server is sent inside client, and available as srv
>   puts "Client starts: #{self.inspect}"
>   received = srv.take                       # The Client takes a message specifically from the server
>   puts "Client received from " \
>        "#{srv.inspect}: #{received}"
>   puts "Client sends to " \
>        "#{srv.inspect}: pong"
>   srv.send 'pong'                           # The client sends a message specifically to the server
> end
> ```
> 
> Who exactly should I `take` from to definitely finalize the script successfully after all the communication have happened, and not disrupt their communication (e.g., not receive in the main ractor server's `"ping"`, instead of the client)? I actually can do `client.take`, but it would obscure the point ("what's that we are trying to take from the client?..") not make it clearer; also, there is a slight chance that the program would exit before the server receives, so it should be `[client, server].each(&:take)`, and in exactly this order.

`[client, server].each(&:take)` seems fine. any problem?

> What I actually need for showcasing examples is some generic `Ractor.tick`, or `Ractor.dispatch_all` with the meaning "let them do what they could", or, maybe `Ractor#wait` with the meaning "wait till Ractor finishes, but not take its value, and not raise ClosedError if it already was", e.g. some more direct analog of `Thread#join`.

`Ractor#join` for documentation or user code?

> Though, in this example neither `#take` nor (imaginary) `#wait` will not do, it requires `Ractor.tick`:
> 
> ```ruby
> r = Ractor.new {
>   puts "Before first receive"
>   v1 = Ractor.receive
>   puts "Received: #{v1}"
>   v2 = receive
>   puts "Received: #{v2}"
> }
> sleep(0.1) # <== HERE
> puts "Still not received"
> r.send('message1')
> sleep(0.1)
> puts "Still received only one"
> r.send('message2')
> sleep(0.1)
> ```
> 
> At the place which I've marked with "HERE", `sleep` demonstrates "Ractor has a chance to run, but it did not yet". `Ractor#take` will not do here, and I am not sure what else will.
> 
> So, I am not sure how to proceed, do you have some ideas?

Only for documentation, for example the comment like "# wait here until step A is over"?

zverok commented:
@ko1 You are perfectly right.
I changed to `#take` wherever it was possible and introduced the "only for documentation" (and explicitly described so) `wait` method for other cases.
I also took the liberty to slightly correct the new `#receive_if` method's docs. Please take a look, does it look more reasonable now?..

marcandre commented on ractor.c:
```diff
@@ -1932,6 +1932,100 @@ ractor_moved_missing(int argc, VALUE *argv, VALUE self)
+/*
+ *  Document-class: Ractor::ClosedError
+ *
+ *  Raised when an attempt is made to take something from the Ractor's outgoing port,
+ *  but it is closed with Ractor#close_outgoing, or to send something to Ractor's
+ *  incoming port, and it is closed with Ractor#close_incoming, or an attempt to
+ *  send/take something with ractor which was already terminated.
```
Not 100% accurate, as you can still read from a non-empty Queue after it has been closed. Also, you can't push to a closed outgoing port, nor read from a non-empty incoming port.  I'd rephrase:

Raised when an attempt is made to send a message to a closed port, or to retrieve a message from a closed and empty port. Ports may be closed explicitly with  Ractor#close_outgoing/close_incoming and are closed implicitly when a Ractor terminates.

marcandre commented on ractor.c:
```diff
@@ -1932,6 +1932,100 @@ ractor_moved_missing(int argc, VALUE *argv, VALUE self)
+ */
+
+// Main docs are in ractor.rb, but without this clause there are weird artifacts
+// it their rendering.
```
^it^in

marcandre commented on ractor.rb:
```diff
@@ -1,34 +1,239 @@
+#     Server received: pong
+#
+# It is said that Ractor receives messages via the <em>incoming port</em>, and sends them
+# to the <em>outgoing port</em>. Either one can be disabled with Ractor.close_incoming and
+# Ractor.close_outgoing respectively.
```
^Ractor.^Ractor# (twice)

marcandre commented on ractor.rb:
```diff
@@ -1,34 +1,239 @@
+#     # I see C
+#     # can not access instance variables of classes/modules from non-main Ractors (RuntimeError)
+#
+# Ractors also can't access constants from the main scope, but only if their values are frozen, or
+# they are classes/modules (but again, not their instance variables):
```
Bit about instance variables not accurate. Frozen isn't 100% accurate either. I'd rephrase:

Ractors can access constants if they are shareable. The main Ractor is the only one that can access non-shareable constants.

marcandre commented on ractor.rb:
```diff
@@ -1,34 +1,239 @@
+#     r.take
+#     # I see C
+#     # can not access instance variables of classes/modules from non-main Ractors (RuntimeError)
+#
```
I'd add a note to check the new `# shareable_constant_value` magic comment and a link to the `syntax/comments` doc.

marcandre commented on ractor.rb:
```diff
@@ -38,38 +243,75 @@ def self.new(*args, name: nil, &block)
-  # Multiplex multiple Ractor communications.
+  # call-seq: Ractor.select(*ractors, [yield_value:, move: false]) -> [ractor or symbol, obj]
+  #
+  # Wait for the first ractor to have something in its outgoing port, reads from this ractor, and
```
^Wait^Waits

marcandre commented on ractor.rb:
```diff
@@ -143,80 +461,268 @@ def self.receive_if &b
-  # utility method
+  # Checks if the object is shareable by ractors.
+  #
+  #     Ractor.shareable?(1)            #=> true -- numbers and other immutable basic values are
```
are *frozen*

marcandre reviewed (commented):
I re-read the whole thing. A few last wording changes, otherwise LGTM

marcandre commented:
@zverok also needs to be rebased.

zverok commented:
> @zverok also needs to be rebased.

OK. I find it interesting what the conflict is: while this PR is being discussed, @ko1 added alternative docs for `make_shareable`. I'll try to merge docs together in the best way possible, but the situation is a bit surprising for me.

marcandre commented:
That's frustrating. Let's merge when ready and we can make further improvements later if need be.

zverok commented:
Rebased, fixed comments.

marcandre commented:
Awesome, thank you 👍
